Tropical polyhedra: Types and basic properties

ooinaruhugh opened the pull request:
Polymake has support for tropical polyhedra, but what is there does not interface so nicely with OSCAR.
This PR serves to add a proper Julia type `TropicalPolyhedron` and provide basic functionality, to fix some design decisions about the data type.

This might also be useful in light of https://github.com/oscar-system/Oscar.jl/pull/3989#pullrequestreview-2217007913.

EDIT: I discussed this PR with @micjoswig and we came up with the following outline.
- Given a set of n points in tropical projective d-space, we want to compute the induced covector decomposition of the tropical projective space.
- For tropical polytopes with only finite coordinates, the covector decomposition yields a polyhedral subdivision of the tropical projective torus.  This is implemented in polymake and should be taken from there.
- With infinite coordinates, we obtain an induced covector decomposition in each stratum (which is a tropical projective torus of suitable dimension).
- For arbitrary point sets, the cells in this decomposition might live in different strata of tropical projective space. In each stratum, this gives a polyhedral complex, which all fit together as a cell complex.
- Thus, a data type for a tropical polytope needs keep track of data (e.g. PolyhedralComplexes) for each stratum.

In terms of data types: try a map which send a set (describing the stratum) to a polyhedral complex.  Beware that polyhedral complexes require the vertices to be numbered consecutively.  Because each stratum only sees a subset of the given points, we need an array and a map for translating indices back and forth.

ooinaruhugh committed:
Add convenience conversions for PosInf and NegInf

ooinaruhugh committed:
Include TropicalPolyhedron.jl

ooinaruhugh committed:
Add the actual TropicalPolyhedron.jl

ooinaruhugh commented:
@Sami-Halaseh

ooinaruhugh committed:
Shorten a function

ooinaruhugh commented:
At the moment, questions that need to be discussed:
- How to handle the bookkeeping accross the strata? (@lkastner Might have an opinion, e.g. the question `Map` vs. indexing strata by bitmasks)
- Regarding the "one polyhedral complex per stratum", vertices need to be renamed. Where to do this, or how to present information on boundary cells?
- More remote question, this is also a welcome addition for polymake. Thus, implement in polymake first and then make accessible from OSCAR, or directly in OSCAR?

fingolfin commented:
What the status of this PR?

ooinaruhugh commented:
This has been pushed to my backlog over the last few months, but I just finished a great deal of work. So thank you for reminding me of this.

At the moment, I just had some prototype code.

micjoswig commented:
Let's first finalize #4061.  Afterwards it will be easier to see how we want to design this.

One aspect to take into account is that @YueRen chose a nonhomogeneous setting.  However, the natural interface between tropical convexity and tropical algebraic geometry goes through tropical hyperplane arrangements described by homogeneous linear forms.

fingolfin commented:
In the meantime #4061 has been merged a few weeks ago. So if @ooinaruhugh is still interested and has some spare time, perhaps this can be brought forward? :-)

micjoswig commented:
The overall goal still stands.

To make progress now, we tackle the following slightly simpler problem. first Input: point configuration in tropical projective space; Output: covector decomposition induced on tropical projective torus (which is an ordinary polyhedral complex).  This can be inherited from polymake directly.

ooinaruhugh committed:
Add covector decompositions and pseudovertices

ooinaruhugh commented:
While testing the `polytope_covector_decomposition`, I noticed inconsistencies regarding what 'pseudovertices' in polymake are. `PSEUDOVERTICES` from polymake gives rays for some reason, and `POLYTOPE_COVECTOR_DECOMPOSITION` does not agree with the visualization of the tropical polytope (from `VISUAL`).
We need to investigate this on the polymake side.

ooinaruhugh committed:
Add exports

ooinaruhugh committed:
Distinguish between trop. polytope and point config.

ooinaruhugh committed:
Fix typo

ooinaruhugh committed:
Add initial round of tests for trop. polytope

ooinaruhugh committed:
Add some tests for covector decompositions

ooinaruhugh committed:
Add test for unbounded tropical polytope

ooinaruhugh committed:
Add some docstrings

benlorenz commented:
> While testing the `polytope_covector_decomposition`, I noticed inconsistencies regarding what 'pseudovertices' in polymake are. `PSEUDOVERTICES` from polymake gives rays for some reason, and `POLYTOPE_COVECTOR_DECOMPOSITION` does not agree with the visualization of the tropical polytope (from `VISUAL`). We need to investigate this on the polymake side.

Can you provide an explicit (polymake) example of this discrepancy?

ooinaruhugh committed:
Fix typo in tests

ooinaruhugh committed:
Backport fix for pseudovertices from polymake

ooinaruhugh committed:
Patch-up covector_decomposition

ooinaruhugh committed:
Add some tests and tentative tests

ooinaruhugh committed:
Add a docstring

ooinaruhugh committed:
Add some docstrings

ooinaruhugh committed:
Fix n_pseudovertices

ooinaruhugh committed:
Diversify test of constructors

ooinaruhugh committed:
Change !=(0) to !iszero

ooinaruhugh committed:
Add examples in docstrings

ooinaruhugh committed:
Add a page for tropical polyhedra in docs

fingolfin commented:
ping @ooinaruhugh

ooinaruhugh commented:
The discussed discrepancies meant that we had to fix some parts in `polymake` regarding tropical polytopes. I'll finish the test cases which reflect how the output of the implemented functions should be mathematically speaking. But we have to wait for `v4.14` of `polymake` in the end. Unless I should implement a temporary workaround that's only going to stay for a short time.

ooinaruhugh committed:
Add broken tests (as of polymake <= 4.13)

ooinaruhugh committed:
Tie broken test to polymake version

ooinaruhugh committed:
Update testcases to new polymake version

ooinaruhugh committed:
Remove some workarounds for old versions of polymake

ooinaruhugh committed:
Add tests for max-tropical polyhedra

ooinaruhugh committed:
Fix last max-tropical tests

ooinaruhugh commented:
I've finally implemented testcases for both min- and max-tropical objects. This just needs a couple of eyes to run over and a tag for the release notes, which I can't add.

joschmitt commented:
> I've finally implemented testcases for both min- and max-tropical objects. This just needs a couple of eyes to run over and a tag for the release notes, which I can't add.

I added the "use title" label (and changed the title). Feel free to change the title to something that fits better or let me know if you want another label.

lgoettgens commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,314 @@
+  ind = findall(1:size(CT, 1)) do i
+    all(!iszero, CT[i,:])
+  end
+  return length(ind)
```
You could use `count` instead of taking `length` of `findall`, avoiding the allocation of the findall result

ooinaruhugh replied:
Thanks, I changed the two occurrences of this pattern

lgoettgens commented on test/TropicalGeometry/tropical_polyhedra.jl:
```diff
@@ -0,0 +1,244 @@
+  P2 = tropical_convex_hull(pts2)
+  @test ambient_dim(P2) == 2
+  @test dim(P2) == 2
+  # the following are unbroken with polymake v4.14
```
Polymake 4.14 has arrived in Oscar in the meantime, so these comments (and maybe also the polymake version checks in this file) are obsolete

micjoswig replied:
> Polymake 4.14 has arrived in Oscar in the meantime, so these comments (and maybe also the polymake version checks in this file) are obsolete

Comments removed; version checks (for test cases) left in place for now.

lgoettgens reviewed (commented):
Two minor things I've seen in a first glance.

One more important point: If I remember correctly, (almost) every field access of a Polymake object should have a type assertion on the julia side to help type inference. Maybe @benlorenz can comment on that

ooinaruhugh committed:
Change `findall|>length` into count

ooinaruhugh committed:
Add some type assertions for polymake objects

ooinaruhugh committed:
Fix typo with `count`

micjoswig committed:
Update tropical_polyhedra.md

trivial typo in docs

micjoswig reviewed (approved):
Looks good to me.

micjoswig committed:
Update tropical_polyhedra.jl

comments concerning polymake version removed: 4.14 is used in OSCAR now.  Version guards for tests left in place for now.

lgoettgens commented on src/exports.jl:
```diff
@@ -1846,3 +1854,4 @@ export young_tableau
 export zero_section
 export zonotope
 export zonotope_vertices_fukuda_matrix
+export n_pseudovertices
```
This is not sorted alphabetically, and it should get an alias to `number_of_pseudovertices` in `src/aliases.jl`

lgoettgens commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,308 @@
+
+Returns the number of vertices of `P`.
+"""
+n_vertices(P::TropicalPolyhedron) = size(pm_object(P).VERTICES::AbstractMatrix, 1)
```
```suggestion
n_vertices(P::TropicalPolyhedron) = nrows(pm_object(P).VERTICES::AbstractMatrix)
```

lgoettgens commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,308 @@
+end
+pseudovertices(P::Union{TropicalPolyhedron{M},TropicalPointConfiguration{M}}) where {M<:MinOrMax} = pseudovertices(PointVector{TropicalSemiringElem{M}}, P)
+
+n_pseudovertices(P::TropicalPointConfiguration) = pm_object(P).PSEUDOVERTICES |> size |> first
```
```suggestion
n_pseudovertices(P::TropicalPointConfiguration) = pm_object(P).PSEUDOVERTICES |> nrows
```

lgoettgens commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,308 @@
+Checks whether `P` is bounded in the tropical projective torus.
+"""
+function is_bounded(P::TropicalPolyhedron)
+  return all(!iszero, pm_object(P).VERTICES::AbstractMatrix)
```
many of these type assertions are not concrete enough to actually help the compiler, thus leading to very poorly inferred code. E.g. look at the following example:
```julia
julia> P = tropical_convex_hull(min, QQ[0 -1 0; 0 -4 -1]);

julia> @inferred is_bounded(P)
ERROR: return type Bool does not match inferred return type Any
```

I don't have experience on how to do this better for polymake matrix types, so it would be best to consult @benlorenz or @lkastner on that

benlorenz replied:
For this example `::Polymake.Matrix{<:Polymake.TropicalNumber}` should work, for rational matrices probably `::Polymake.Matrix{Polymake.Rational}`.

The reason for all this is that there is no type information on these properties and they are unwrapped from generic pointers at runtime.

lgoettgens commented on test/TropicalGeometry/tropical_polyhedra.jl:
```diff
@@ -0,0 +1,242 @@
+__sign = minOrMax == min ? 1 : -1
+oo = minOrMax == min ? inf : -inf
+
+@testset "TropicalPolyhedron{$minOrMax}" begin
```
```suggestion
@testset "TropicalPolyhedron{$minOrMax}" for (minOrMax, __sign, oo) in [(min, 1, inf), (max, -1, -inf)]
```
some suggestion to make this more readable

lgoettgens commented on test/TropicalGeometry/tropical_polyhedra.jl:
```diff
@@ -0,0 +1,242 @@
+      IncidenceMatrix([[1],[2],[3]])
+    ]) broken=polymake_version <= v"400.1400.0+0"
+end
+end
```
```suggestion
```
due to merging of for loop and testset above

lgoettgens commented on test/TropicalGeometry/tropical_polyhedra.jl:
```diff
@@ -0,0 +1,242 @@
+  @test issetequal(pseudovertices(P2), [TT.(__sign*[0,1,2]),TT.(__sign*[0,-1,2])]) 
+  @test n_pseudovertices(P2) == 2 
+  @test maximal_covectors(P2) |> length == 1
+  @test maximal_covectors(P2) |> first == IncidenceMatrix([[1], [2], [3]]) broken=polymake_version <= v"400.1400.0+0"
```
reading the `broken` condition again, this states that the test in broken with *the initial yggdrasil build* of polymake 4.14, and all previous versions. The problem with that is that some rebuild of polymake without any update to polymake itself (e.g. due to a new FLINT release), would increase this version number and thus changing the output on the condition.

Is the bug already fixed with polymake 4.14 (and thus polymake_jll 400.1400.0+0), or will the fix only be in the next polymake release after that? Depending on that, I could make a suggetsion on how to change the condition

ooinaruhugh replied:
There's a hotfix for the specific thing that breaks the test, so it would still be v4.14.
The last time I spoke with @benlorenz, this was the specific condition I should test for.

benlorenz replied:
There was some rework for 4.14 which fixed many things but there are some things (that unfortunately lacked proper tests) that are still broken / not yet adapted to the new changes. We wanted to fix this in a patch-release for 4.14, exact versioning of the jll still to be decided upon. We already have a hotfix branch for polymake but the fix seems not yet complete. Kamillo will investigate it next week.

fingolfin replied:
What's the status of this hotfix?

YueRen commented on docs/src/TropicalGeometry/tropical_polyhedra.md:
```diff
@@ -0,0 +1,68 @@
+    which is a polyhedral subcomplex of the covector decomposition for the corresponding `TropicalPointConfiguration`.
+- Currently, OSCAR supports
+  - point sets $V$ with entirely finite coordinates
+  - point sets $V$ with infinite coordinates and $\mathrm{tconv}(V)$ inside the tropical projective torus
```
Quick question regarding these two bullet points.  In case $V$ has "entirely finite coordinates", is $\mathrm{tconv}(V)$ not inside the tropical projective torus? Or why is the "$\mathrm{tconv}(V)$ inside the tropical projective torus" omited in the first bullet point?

In case $\mathrm{tconv}(V)$ is always regarded as a subset of the tropical projective torus, I'd just say that $V$ can be any finite set in $\mathbb{T}^d$ but that $\mathrm{tconv}(V)$ is always regarded as a of $\mathbb{TP}^{d-1}$.

ooinaruhugh replied:
The difference is between 'tropical projective torus' and 'tropical projective space'.
If the coordinates are finite, then $\mathrm{tconv}(V)$ is in the t.p. torus, if there are infinite coordinates in $V$,
then $\mathrm{tconv}(V)$ is possibly unbounded. Then, you could possibly look at the points with infinite coordinates, in the t.p. space, but this is not implemented in this PR. 

This text is supposed to mean: "In the inifinite coordinates case, we can do $\mathrm{tconv}(V)\cap\mathbb{R}^d/\mathbb{R}\mathbf{1}$ right now"

YueRen commented on docs/src/TropicalGeometry/tropical_polyhedra.md:
```diff
@@ -0,0 +1,68 @@
+
+## Constructors
+Both `TropicalPolyhedron` and `TropicalPointConfiguration` can be constructed from the following data:
+- a rational matrix and the tropical addition
```
```suggestion
- a rational matrix and specifying min- or max-convention
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
````diff
@@ -0,0 +1,151 @@
+Max tropical polyhedron in tropical projective torus of dimension 2
+```
+"""
+tropical_convex_hull(convention::MinOrMax, V::QQMatrix) = _tropical_convex_hull(TropicalPolyhedron, convention, V)
````
```suggestion
tropical_convex_hull(V::QQMatrix, convention::MinOrMax=min) = _tropical_convex_hull(TropicalPolyhedron, convention, V)
```

Can you please put `convention` last, make it optional with default min?  

I think there might an unfortunate clash of conventions between the polyhedral part and the tropical part of OSCAR: input which specifies the return type is listed first in the polyhedral part while conventions and tropical semiring maps are listed last in the tropical part (because I'm used to C where optional input has to be last).

YueRen replied:
For the sake of consistency, can you please change all user-facing functions accordingly?  (I'm fine with the non-user facing functions staying the way they are)

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,308 @@
+"""
+function covector_decomposition(P::TropicalPolyhedron; dehomogenize_by=1)
+  if !isnothing(dehomogenize_by)
+    return Polymake.tropical.polytope_subdivision_as_complex(P.pm_tpolytope, dehomogenize_by-1)::Polymake.BigObject |> polyhedral_complex
```
```suggestion
    return Polymake.tropical.polytope_subdivision_as_complex(pm_object(P), dehomogenize_by-1)::Polymake.BigObject |> polyhedral_complex
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,308 @@
+@doc raw"""
+    vertices([as::Type{T} = PointVector,] P::TropicalPolyhedron)
+
+Returns an iterator over the vertices of `P`, that is, a minimal generating set `V`
```
```suggestion
Return an iterator over the vertices of `P`, that is, a minimal generating set `V`
```

The convention is not to use third person singular (don't ask me why).

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,308 @@
+@doc raw"""
+    n_vertices(P::TropicalPolyhedron)
+
+Returns the number of vertices of `P`.
```
```suggestion
Return the number of vertices of `P`.
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,308 @@
+@doc raw"""
+    pseudovertices([as::Type{T} = PointVector,] P::TropicalPolyhedron)
+
+Returns an iterator over the pseudovertices of `P` in the tropical projective torus, 
```
```suggestion
Return an iterator over the pseudovertices of `P` in the tropical projective torus, 
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,308 @@
+@doc raw"""
+    maximal_covectors([as::Type{T} = PointVector,] P::TropicalPointConfiguration)
+
+Returns an iterator over the maximal covectors with respect to `P`.
```
```suggestion
Return an iterator over the maximal covectors with respect to `P`.
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,308 @@
+@doc raw"""
+    maximal_covectors([as::Type{T} = PointVector,] P::TropicalPolyhedron)
+
+Returns an iterator over the maximal covectors of `P`.
```
```suggestion
Return an iterator over the maximal covectors of `P`.
```

YueRen reviewed (changes requested):
Some quick things I caught, but everything looks really solid over all, good job!

The only major change is to put convention last as input argument, for the sake of consistency with the rest of the tropical package.

ooinaruhugh committed:
Add alias for n_pseudovertices

ooinaruhugh committed:
Replace `size(...,1)` with `nrows`

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Fix typo in doctest

ooinaruhugh committed:
Refine type assertions

ooinaruhugh committed:
Rewrite for loop in tests

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Fix grammar in docstrings

Co-authored-by: Yue Ren <yue.ren.kl@gmail.com>

ooinaruhugh committed:
Replace `.pm_tpolytope` acess with `pm_object(...)'

Co-authored-by: Yue Ren <yue.ren.kl@gmail.com>

ooinaruhugh committed:
Move `convention` optional argument to end of arguments

Co-authored-by: Yue Ren <yue.ren.kl@gmail.com>

ooinaruhugh committed:
Add missing docstrings

ooinaruhugh committed:
Adjust test to changed function definitions

ooinaruhugh committed:
Change wording in docs

Co-authored-by: Yue Ren <yue.ren.kl@gmail.com>

ooinaruhugh committed:
Fix typos in tests, make ready for new Polymake release

ooinaruhugh committed:
Fix duplicate in docs

fingolfin commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+      TropicalPolyhedron{M}(p::Polymake.BigObject) where M<:MinOrMax
+
+  Constructs a `TropicalPolyhedron` corresponding to a `Polymake.BigObject` with tropical addition `M`.
+  """
```
We should not document this constructor (it's internal). Instead there should be a docstring on the `struct` definition itself (so that users can find out what this struct is doing, and ideally also what the proper way for constructing an instance is: e.g. by proving a code example that use suitable functions (`tropical_convex_hull` I am guessing would be one?), or with `@ref`

fingolfin commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+  """
+  TropicalPolyhedron{M}(p::Polymake.BigObject) where {M<:MinOrMax} = new{M}(p)
+end
+struct TropicalPointConfiguration{M<:MinOrMax} <: PolyhedralObject{QQFieldElem}
```
```suggestion

struct TropicalPointConfiguration{M<:MinOrMax} <: PolyhedralObject{QQFieldElem}
```

fingolfin commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+      TropicalPointConfiguration{M}(p::Polymake.BigObject) where M<:MinOrMax
+
+  Constructs a `TropicalPointConfiguration` corresponding to a `Polymake.BigObject` with tropical addition `M`.
+  """
```
same comment as above applies here as well.

fingolfin commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+
+  return TropicalPolyhedron{M}(p)
+end
+function tropical_polyhedron(A::QQMatrix, B::QQMatrix, convention::MinOrMax=min)
```
```suggestion

function tropical_polyhedron(A::QQMatrix, B::QQMatrix, convention::MinOrMax=min)
```

fingolfin commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,368 @@
+end
+pseudovertices(P::Union{TropicalPolyhedron{M},TropicalPointConfiguration{M}}) where {M<:MinOrMax} = pseudovertices(PointVector{TropicalSemiringElem{M}}, P)
+
+n_pseudovertices(P::TropicalPointConfiguration) = pm_object(P).PSEUDOVERTICES |> nrows
```
This is not meant as a blocker to this PR, it could be removed in a follow-up, but I want to record this point of view here:

I'd greatly prefer if this code did not use the pipeline operator `|>` -- it's just yet another Julia feature that's cool if you know it (I personally like it!), and it's great for the REPL -- but many people writing Julia code will still not be familiar with it, raising the barrier for working on this part of the code.

In fact I just checked, and this operator is used just 4 times in all of Oscar.jl -- twice in `experimental`, and twice in `src`.

fingolfin commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
````diff
@@ -0,0 +1,368 @@
+2
+```
+"""
+n_points(P::TropicalPointConfiguration) = pm_object(P).POINTS::Polymake.Matrix{<:Polymake.TropicalNumber} |> size |> first
````
Untested, but, doesn't this also work:
```suggestion
n_points(P::TropicalPointConfiguration) = pm_object(P).POINTS::Polymake.Matrix{<:Polymake.TropicalNumber} |> nrows
```
or even better
```suggestion
n_points(P::TropicalPointConfiguration) = nrows(pm_object(P).POINTS::Polymake.Matrix{<:Polymake.TropicalNumber})
```

fingolfin commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,368 @@
+   pv = pm_object(P).PSEUDOVERTICES::Polymake.Matrix{Polymake.Rational}
+   cov = pm_object(P).DOME.TROPICAL_SPAN_MAXIMAL_COVECTOR_CELLS::IncidenceMatrix
+   ct = pm_object(P).PSEUDOVERTEX_COARSE_COVECTORS::Polymake.Matrix{<:Integer}
+   ind = findall(1:size(ct, 1)) do i
```
```suggestion
   ind = findall(1:nrows(ct)) do i
```

fingolfin commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,368 @@
+    print(io, "Max tropical polyhedron in tropical projective torus of dimension $d")
+  end
+end
+function Base.show(io::IO, P::TropicalPointConfiguration{M}) where {M<:MinOrMax}
```
```suggestion

function Base.show(io::IO, P::TropicalPointConfiguration{M}) where {M<:MinOrMax}
```

fingolfin reviewed (commented):
Would be nice to get this merged. There is a bunch of minor code review comments that could be addressed with a few clicks of "apply suggestions". A couple others seem to be resolved and just need to be marked as such.

There is also the issue of the "polymake hotfix", no idea what the status there is?

benlorenz commented:
> There is also the issue of the "polymake hotfix", no idea what the status there is?

It was released as part of polymake 4.15 and is available as polymake_jll-400.1500.000. This version is already in use by Oscar.
Strictly speaking we could either add a compat entry for `polymake_jll` to enforce this, or set `broken= pmversion < v"400.1500.0"` but I don't think this is necessary since there is nothing else that could block this update.

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+  @doc raw"""
+      TropicalPolyhedron{M}(p::Polymake.BigObject) where M<:MinOrMax
+
+  Constructs a `TropicalPolyhedron` corresponding to a `Polymake.BigObject` with tropical addition `M`.
```
```suggestion
  Construct a `TropicalPolyhedron` corresponding to a `Polymake.BigObject` with tropical addition `M`.
```

Consistency with other parts of OSCAR code.

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+  @doc raw"""
+      TropicalPointConfiguration{M}(p::Polymake.BigObject) where M<:MinOrMax
+
+  Constructs a `TropicalPointConfiguration` corresponding to a `Polymake.BigObject` with tropical addition `M`.
```
```suggestion
  Construct a `TropicalPointConfiguration` corresponding to a `Polymake.BigObject` with tropical addition `M`.
```

Consistency with other parts of OSCAR code.

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
````diff
@@ -0,0 +1,151 @@
+```jldoctest tropical_convex_hull
+julia> TT = tropical_semiring(max);
+
+julia> A = TT.(identity_matrix(QQ, 3));
````
```suggestion
julia> T = tropical_semiring(max);

julia> A = T.(identity_matrix(QQ, 3));
```

Consistency with other tropical docu.

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
````diff
@@ -0,0 +1,151 @@
+
+The vertices of a tropical convex hull may also be given by the rows of a tropical matrix:
+```jldoctest tropical_convex_hull
+julia> A = TT[0 -4 -1; 0 -1 0];
````
```suggestion
julia> A = T[0 -4 -1; 0 -1 0];
```

Consistency with other tropical docu.

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
````diff
@@ -0,0 +1,151 @@
+```jldoctest tropical_point_configuration
+julia> TT = tropical_semiring(max);
+
+julia> A = TT.(identity_matrix(QQ, 3));
````
```suggestion
julia> T = tropical_semiring(max);

julia> A = T.(identity_matrix(QQ, 3));
```

Consistency with other tropical docu.

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
````diff
@@ -0,0 +1,151 @@
+```
+The tropical point configuration may also be given by the rows of a tropical matrix:
+```jldoctest tropical_point_configuration
+julia> A = TT[0 -4 -1; 0 -1 0];
````
```suggestion
julia> A = T[0 -4 -1; 0 -1 0];
```

Consistency with other tropical docu.

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+@doc raw"""
+    tropical_point_configuration(V::QQMatrix[, convention::MinOrMax=min])
+
+Constructs the tropical point configuration of the vertices `V` in the tropical projective torus
```
```suggestion
Construct the tropical point configuration of the vertices `V` in the tropical projective torus
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+@doc raw"""
+    tropical_convex_hull(V::QQMatrix[,convention::MinOrMax=min])
+
+Constructs the tropical convex hull of the vertices `V` in the tropical projective torus
```
```suggestion
Construct the tropical convex hull of the vertices `V` in the tropical projective torus
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+@doc raw"""
+    tropical_point_configuration(V::MatElem{TropicalSemiringElem{M}}) where M<:MinOrMax
+
+Constructs the tropical point configuration of the vertices `V` in the tropical projective torus
```
```suggestion
Construct the tropical point configuration of the vertices `V` in the tropical projective torus
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+@doc raw"""
+    tropical_convex_hull(V::MatElem{TropicalSemiringElem{M}}) where M<:MinOrMax
+
+Constructs the tropical convex hull of the vertices `V` in the tropical projective torus
```
```suggestion
Construct the tropical convex hull of the vertices `V` in the tropical projective torus
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+@doc raw"""
+    tropical_polyhedron(A::MatElem{TropicalSemiringElem{M}}, B::MatElem{TropicalSemiringElem{M}}) where M<:MinOrMax
+
+Constructs a `TropicalPolyhedron` as the set of points in the tropical projective torus satisfying the inequalities $A\odot x\leq B\odot x$
```
```suggestion
Construct a `TropicalPolyhedron` as the set of points in the tropical projective torus satisfying the inequalities $A\odot x\leq B\odot x$
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+with `convention` as tropical addition.
+
+# Examples
+Defines a point configuration containing a tropical unit ball in the 2-dimensional tropical projective torus with `max` as addition:
```
```suggestion
Define a point configuration containing a tropical unit ball in the 2-dimensional tropical projective torus with `max` as addition:
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/constructors.jl:
```diff
@@ -0,0 +1,151 @@
+with `convention` as tropical addition.
+
+# Examples
+Defines a point configuration containing a tropical unit ball in the 2-dimensional tropical projective torus with `max` as addition:
```
```suggestion
Define a point configuration containing a tropical unit ball in the 2-dimensional tropical projective torus with `max` as addition:
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,368 @@
+
+julia> TT = tropical_semiring();
+
+julia> Q = tropical_convex_hull(TT[0 1 2; inf 0 3; inf inf 0; 0 1 0]);
```
```suggestion
julia> T = tropical_semiring();

julia> Q = tropical_convex_hull(T[0 1 2; inf 0 3; inf inf 0; 0 1 0]);
```

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,368 @@
+
+  return SubObjectIterator{as}(
+    P,_pseudovertices,n
+  )
```
```suggestion
  return SubObjectIterator{as}(P, _pseudovertices, n)
```

To be consistent with code below.

YueRen commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,368 @@
+function pseudovertices(as::Type{PointVector{T}}, P::TropicalPolyhedron) where {T<:TropicalSemiringElem}
+  CT = pm_object(P).PSEUDOVERTEX_COARSE_COVECTORS::Polymake.Matrix{<:Integer}
+  n = count(!iszero, eachrow(CT))
+  TT = tropical_semiring(convention(P))
```
```suggestion
```

This line appears to be unnecessary.

YueRen reviewed (changes requested):
@ooinaruhugh Looks great, thanks for your contribution!  Here just some very minor nitpicks.

ooinaruhugh committed:
Typography, whitespace and grammatical consistency

Co-authored-by: Max Horn <max@quendi.de>
Co-authored-by: Yue Ren <yue.ren.kl@gmail.com>

ooinaruhugh committed:
Move docstrings

ooinaruhugh committed:
Remove reference to broken polymake version

fingolfin reviewed (approved):
Looks good to me, thank you very much!

benlorenz commented on src/TropicalGeometry/TropicalPolyhedron/properties.jl:
```diff
@@ -0,0 +1,366 @@
+end
+
+@doc raw"""
+    n_vertices(P::TropicalPointConfiguration)
```
```suggestion
    n_points(P::TropicalPointConfiguration)
```
(according to the doctest and code below)

ooinaruhugh committed:
Update properties.jl

Co-authored-by: Benjamin Lorenz <benlorenz@users.noreply.github.com>

YueRen reviewed (approved):
Strong approve.
